src/mainwindow.cpp

```diff
@@ -193,12 +198,35 @@ void MainWindow::openImageFile(QString fileName) {
     fm_path_unref(path);
     return;
   }
-  // load the image file asynchronously
-  loadImage(path);
-  loadFolder(fm_path_get_parent(path));
+  if (QFileInfo(fileName).isDir()) {
+    if(fm_path_equal(path, folderPath_) || !dirHasImage(fileName)) {
+      fm_path_unref(path);
+      return;
+    }
+    if(currentFile_)
+      fm_path_unref(currentFile_);
+    currentFile_ = nullptr;
+    loadFolder(path);
+  } else {
+    // load the image file asynchronously
+    loadImage(path);
+    loadFolder(fm_path_get_parent(path));
+  }
   fm_path_unref(path);
 }
 
+bool MainWindow::dirHasImage(QString dirname)
+{
+  QList<QByteArray> formats = QImageReader::supportedImageFormats();
+  QStringList formatsFilters;
+  for (const QByteArray& format: formats)
+    formatsFilters << QString("*.") + format;
+  QDir dir(dirname);
+  dir.setNameFilters(formatsFilters);
+  dir.setFilter(QDir::Files | QDir::NoDotAndDotDot);
+  return !dir.entryList().isEmpty();
+}
+
```

Comment:
MainWindow::dirHasImage() seems redundant to me because it doesn't use any class member. We can remove it (from the header too) and put its contents into openImageFile() as:
````
void MainWindow::openImageFile(QString fileName) {
  FmPath* path = fm_path_new_for_str(qPrintable(fileName));
  if(currentFile_ && fm_path_equal(currentFile_, path)) {
    // the same file! do not load it again
    fm_path_unref(path);
    return;
  }
  if (QFileInfo(fileName).isDir()) {
    if(fm_path_equal(path, folderPath_)) {
      fm_path_unref(path);
      return;
    }
    QList<QByteArray> formatsBytes = QImageReader::supportedImageFormats();
    QStringList formats;
    for (const QByteArray& formatBytes: formatsBytes)
      formats << QString("*.") + formatBytes;
    QDir dir(fileName);
    dir.setNameFilters(formats);
    dir.setFilter(QDir::Files | QDir::NoDotAndDotDot);
    if(dir.entryInfoList().isEmpty()) {
      fm_path_unref(path);
      return;
    }
    if(currentFile_)
      fm_path_unref(currentFile_);
    currentFile_ = nullptr;
    loadFolder(path);
  } else {
    // load the image file asynchronously
    loadImage(path);
    loadFolder(fm_path_get_parent(path));
  }
  fm_path_unref(path);
}
````
Please also correct the indentation (double space) as above and this PR will be ready for a merge.